### ISSUE TITLE
defFutureAny returns {value: Object}

### ISSUE THREAD

**Issue by colinbes** [joescii/lift-ng#29]
Not sure if I am missing something but I have changed some code to start using defFutureAny as opposed to defAny due to nature of source of data.

Since making this change instead, in my angular controller the response is not as expected. Prior to change I would receive return data `[Object, Object, Object, Object, Object, Object, Object]`, now with Future I receive `Object {value: Array(7)}` and am not sure why or where the 'value' field comes from.

In both before and after change my JValue result is `Full(JArray(List(JObject(List(JField(_id,JObject(List(JField(...`

Code snippet is
```
  def getAllCachedRecipes(): Future[Box[JValue]] = { 
    DB.programs.getCachedProgramsByTag("").map(f =>  f match {        
      case Left(err) => {
        val errors: String = err.map(e => e.error+" "+e.message).mkString("\n")
				Failure(errors)
      }
      
      case Right(programs) => {
      	try {
      		val res = Full(Extraction.decompose(programs))
      		println(s"res $res")
      		res
      	} catch {
      		case ex:Exception => {
      			Failure(ex.getMessage())
      		}
      	}
      }
    }) 
  }
```
I am using `import net.liftmodules.ng.FutureConversions._` to handle conversion of Scala's Future to  LiftWeb's LAFuture - code snippet below is from LiftWeb snippet
```
....jsObjFactory().defFutureAny("getAllRecipes", RecipeService.getAllCachedRecipes())
```

Easily fixed in the controller by access return value's value field but as the field is unexpected I feel uncomfortable doing this.

I also just updated to ng-lift 0.9.5 and it behaves the same.

Suggestions/ideas?

**Comment by joescii** [comment]
I took a glance around the source, and I don't have an immediate answer. I likewise feel that it should not be different in the client. The intention of having `defAny` and `defFutureAny` is to allow different data types on the server to resolve to the same form on the client.

Let me know if you have time to whip together a sample project with the same logic served out using the two functions? I have several projects that may be easy starting points [here](https://github.com/joescii?utf8=%E2%9C%93&tab=repositories&q=lift-ng&type=&language=) as well as the `test-project` here in this repo.

**Comment by colinbes** [comment]
It's the least I can do. Will also help to see if issue repeats itself.

**Comment by joescii** [comment]
Cool, let me know if you have any problems. That'll help me get a head start on it this weekend.

**Comment by colinbes** [comment]
Joe, I have placed mini project showing same behavior at https://github.com/colinbes/ng-lift-test

Note, I haven't tried replicating issue with non-array data. 

Cheers
Colin

**Comment by joescii** [comment]
So your example project shows how the `value` field is introduced by changing the function call?

**Comment by colinbes** [comment]
Correct. In DeviceSnippet I define two services
```
val f1: JsObjFactory = jsObjFactory()
        .defAny("getData", getTestData())
        .defFutureAny("getCachedData", getCachedTestData().la)
      
      def services() = angular.module("bc.services").factory("deviceService", f1)
```
where 

```
  def getTestData(): Box[JValue] =  {
    val data = List(TestData("a", 1), TestData("b", 2))
    val res = Full(Extraction.decompose(data))
    println(s"res $res")
    res
  }
  
  def getCachedTestData(): Future[Box[JValue]] =  {
    Future {
      val data = List(TestData("aa", 11), TestData("bb", 22))
      val res = Full(Extraction.decompose(data))
      println(s"futured res $res")
      res
    }
  }
```

In src/main/webapp/static/rel-010/js/controller.js I call these two services and place into scope (I also simply log them to console

```
		deviceService.getData().then(function(data) {
			console.log("received", data)
			$scope.data = data
		})

		deviceService.getCachedData().then(function(data) {
			console.log("received cache", data)
			$scope.futureData = data
		})
```

**Comment by colinbes** [comment]
Let me know if I can assist in any way.

**Comment by joescii** [comment]
Thanks for the reminder to look at this for you. Been busy lately, but don't be shy about coaxing me back over here.

I don't know much about Maven in order to quickly fix my issue. Perhaps you have jetty as a plugin installed system-wide and I don't?

```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 10.869 s
[INFO] Finished at: 2017-04-24T14:17:15-05:00
[INFO] Final Memory: 9M/245M
[INFO] ------------------------------------------------------------------------
[ERROR] No plugin found for prefix 'jetty' in the current project and in the plugin groups [org.apache.maven.plugins, org.codehaus.mojo] available from the repositories [local (/Users/joescii/.m2/repository), central (http://138.197.213.98:8081/artifactory/plugins-release), snapshots (http://138.197.213.98:8081/artifactory/plugins-snapshot), scala-tools.releases (https://oss.sonatype.org/content/repositories/releases/)] -> [Help 1]
```

**Comment by colinbes** [comment]
Hmm, I wonder if it's because of reference to artifactory which is a private repo. I have removed the repository and plugin references to this. 

mvn clean
man jetty:run works for me, loading jetty via maven plugin.

Please pull changes and let me know if this works (I would not think my artifactory reference would be an issue).

**Comment by joescii** [comment]
Ha... that was it! And just in time I have to get interrupted. I'll try to look more tonight.

**Comment by colinbes** [comment]
Did not expect that, I thought would just fail and move on to next repo - live and learn.

**Comment by colinbes** [comment]
I have updated  https://github.com/colinbes/ng-lift-test to show different (unexpected) return types for Future(Failure(...)). 

index.html is updated to show results.

Could I be processing service call in controller incorrectly?

```
deviceService.getFutureFail().then(function(data) {
	console.log("Oops, received future", data)
	$scope.oops2 = data
}, function(err) {
	$scope.futureFailData = err
})
```

**Comment by joescii** [comment]
Again, I apologize for the long delay for me to get back to you with this. The funny thing is it took me about 5 minutes to figure out your problem because late last week a developer for one of my clients had the same problem, lol. 

Short answer, change your type from `Future[Box[JValue]]` to `Future[JValue]` like I did [here](https://github.com/joescii/ng-lift-test/commit/5b2bd2840220866d0e75f5af28bd5083ed8d587a).

Longer answer: when I first started lift-ng Lift was still on 2.5.x. Back then we supported Scala versions prior to 2.11.x, which predates Scala's `Future` support. Hence I had to go the `LAFuture` route with the lift-ng API and provide the conversions as an add-on for 2.11.x and up. Furthermore, I elected to go with the conversions as `Future[T] => LAFuture[Box[T]]` because a Scala `Future` can fail, but a Lift `LAFuture` cannot. Hence for me to conserve the semantics of Scala `Future`, it was necessary to `Box` the inner type. Be on the lookout for future versions of lift-ng where I will drop support for older versions of Scala and Lift, and update the API to use Scala `Future` natively.

A side note: You don't have to use the `JValue` type and serialize the values like you do [here](https://github.com/joescii/ng-lift-test/blob/5b2bd2840220866d0e75f5af28bd5083ed8d587a/src/main/scala/com/besterdesigns/snippet/DeviceSnippet.scala#L54) if you create a case class which extends `NgModel`.
